egs/librispeech/ASR/pruned_transducer_stateless2/scaling.py

```diff
@@ -111,6 +111,77 @@ def backward(
         return x_grad - neg_delta_grad, None, None, None, None, None, None
 
 
+class GradientClipFunction(torch.autograd.Function):
+    @staticmethod
+    def forward(
+        ctx,
+        x: Tensor,
+        grad_norm_threshold: float,  # e.g., 10.0
+        grad_scale_factor: float,  # e.g., 0.9
+        grad_max_norm: float,  # e.g., 5.0
+    ) -> Tensor:
+        if x.requires_grad:
+            ctx.grad_norm_threshold = grad_norm_threshold
+            ctx.grad_scale_factor = grad_scale_factor
+            ctx.grad_max_norm = grad_max_norm
+        return x
+
+    @staticmethod
+    def backward(ctx, x_grad: Tensor) -> Tuple[Tensor, None]:
+        if x_grad.norm() > ctx.grad_norm_threshold:
+            # Zero gradients if its norm is larger than the threshold
+            x_grad = torch.zeros_like(x_grad)
+        else:
+            # Scale down gradient
+            x_grad = x_grad * ctx.grad_scale_factor
+
+            # Limit gradient norm to a maxmimum value
+            clip_coef = ctx.grad_max_norm / (x_grad.norm() + 1e-6)
```

Comment:
here, instead of x_grad.norm(), need `x_grad.norm(dim=ctx.channel_dim)`.
Also, cannot apply grad_norm_threshold using an if statement like this, would need to apply it as some kind
of multiplicative mask.  But it might be easier to just not apply it for now.